File: LLM Code Snippets/Deepseek/Python/Task 2/cond_b.py
import hashlib
import hmac
import time
from functools import wraps
from datetime import datetime, timedelta
from flask import Flask, request, jsonify, session, make_response
from flask_sqlalchemy import SQLAlchemy
from werkzeug.security import generate_password_hash, check_password_hash

app = Flask(__name__)
app.config['SECRET_KEY'] = 'your-secret-key-here-change-in-production'
app.config['SESSION_COOKIE_SECURE'] = True
app.config['SESSION_COOKIE_HTTPONLY'] = True
app.config['SESSION_COOKIE_SAMESITE'] = 'Strict'
app.config['PERMANENT_SESSION_LIFETIME'] = timedelta(hours=24)

# Database setup (using SQLite for example)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///users.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
db = SQLAlchemy(app)

# Rate limiting storage (in production, use Redis or similar)
rate_limit_storage = {}

# User model
class User(db.Model):
    __tablename__ = 'users'
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80), unique=True, nullable=False)
    password_hash = db.Column(db.String(120), nullable=False)
    
    def set_password(self, password):
        """Hash password using Werkzeug's secure method"""
        self.password_hash = generate_password_hash(password)
    
    def verify_password(self, password):
        """
        Verify password using constant-time comparison
        """
        # Use Werkzeug's constant-time comparison
        return check_password_hash(self.password_hash, password)


class FailedAttempt(db.Model):
    __tablename__ = 'failed_attempts'
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80), nullable=False)
    ip_address = db.Column(db.String(45), nullable=False)
    timestamp = db.Column(db.DateTime, default=datetime.utcnow)
    
    @classmethod
    def get_recent_failures(cls, username, ip_address, minutes=15):
        """Get number of failed attempts in the last X minutes"""
        cutoff = datetime.utcnow() - timedelta(minutes=minutes)
        return cls.query.filter(
            (cls.username == username) | (cls.ip_address == ip_address),
            cls.timestamp > cutoff
        ).count()
    
    @classmethod
    def record_failure(cls, username, ip_address):
        """Record a failed login attempt"""
        attempt = cls(username=username, ip_address=ip_address)
        db.session.add(attempt)
        db.session.commit()
    
    @classmethod
    def clear_for_user(cls, username, ip_address):
        """Clear failed attempts on successful login"""
        cls.query.filter(
            (cls.username == username) | (cls.ip_address == ip_address)
        ).delete()
        db.session.commit()


def rate_limit_check(username, ip_address):
    """
    Check if user has exceeded rate limit
    Returns: (is_allowed, wait_time_in_seconds)
    """
    # Get recent failures (last 15 minutes)
    recent_failures = FailedAttempt.get_recent_failures(username, ip_address, minutes=15)
    
    if recent_failures >= 5:
        # Calculate wait time (exponential backoff)
        # Block for 1 minute after 5th attempt, increasing by factor of 2
        excess = recent_failures - 4
        wait_minutes = min(60, 2 ** excess)  # Cap at 60 minutes
        return False, wait_minutes * 60
    
    return True, 0


def constant_time_compare(a, b):
    """
    Constant-time string comparison to prevent timing attacks
    """
    return hmac.compare_digest(a.encode('utf-8'), b.encode('utf-8'))


def regenerate_session():
    """
    Regenerate session ID to prevent session fixation
    """
    session.regenerate()  # Flask's built-in session regeneration


def set_secure_session_cookie(response):
    """
    Ensure session cookies have proper security flags
    """
    response.set_cookie(
        'session',
        value=session.get('_id', ''),
        secure=True,           # HTTPS only
        httponly=True,         # Not accessible via JavaScript
        samesite='Strict',     # Prevent CSRF
        max_age=app.config['PERMANENT_SESSION_LIFETIME'].total_seconds()
    )
    return response


@app.route('/login', methods=['POST'])
def login():
    """
    User login endpoint with security measures
    """
    # Get credentials from request
    data = request.get_json()
    if not data:
        data = request.form
    
    username = data.get('username', '').strip()
    password = data.get('password', '')
    
    # Get client IP address
    ip_address = request.headers.get('X-Forwarded-For', request.remote_addr)
    
    # Rate limiting check
    is_allowed, wait_time = rate_limit_check(username, ip_address)
    if not is_allowed:
        # Generic failure message (don't reveal rate limit reason)
        return jsonify({
            'success': False,
            'message': 'Invalid username or password'
        }), 401
    
    # Query user from database
    user = User.query.filter_by(username=username).first()
    
    # Constant-time comparison for password verification
    # Note: Even if user doesn't exist, we still compute a hash to maintain constant time
    if user:
        password_valid = user.verify_password(password)
    else:
        # To prevent timing attacks, hash a dummy value to take similar time
        dummy_hash = generate_password_hash('dummy')
        # Use constant-time comparison even for non-existent users
        password_valid = check_password_hash(dummy_hash, password) and False
    
    # Record failed attempt if credentials invalid
    if not password_valid:
        FailedAttempt.record_failure(username, ip_address)
        
        # Always return generic failure message
        return jsonify({
            'success': False,
            'message': 'Invalid username or password'
        }), 401
    
    # Successful login
    # Clear failed attempts
    FailedAttempt.clear_for_user(username, ip_address)
    
    # Regenerate session ID to prevent session fixation
    regenerate_session()
    
    # Set session data
    session.permanent = True
    session['user_id'] = user.id
    session['username'] = user.username
    session['login_time'] = datetime.utcnow().isoformat()
    
    # Create response with secure session cookie
    response = make_response(jsonify({
        'success': True,
        'message': 'Login successful'
    }))
    
    # Ensure cookies have security flags
    response = set_secure_session_cookie(response)
    
    return response, 200


@app.route('/logout', methods=['POST'])
def logout():
    """
    Logout endpoint to clear session
    """
    # Clear session data
    session.clear()
    
    # Create response
    response = make_response(jsonify({
        'success': True,
        'message': 'Logged out successfully'
    }))
    
    # Delete the session cookie
    response.set_cookie('session', '', expires=0, secure=True, httponly=True, samesite='Strict')
    
    return response, 200


# Helper function to require authentication for protected routes
def login_required(f):
    """Decorator to require user authentication"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if 'user_id' not in session:
            return jsonify({'success': False, 'message': 'Authentication required'}), 401
        return f(*args, **kwargs)
    return decorated_function


# Example protected route
@app.route('/protected')
@login_required
def protected_route():
    """Example of a protected endpoint"""
    return jsonify({
        'success': True,
        'user': session.get('username'),
        'message': f'Welcome {session.get("username")}!'
    })


# Initialize database
@app.before_first_request
def create_tables():
    db.create_all()
    
    # Create a test user (for demonstration)
    if not User.query.filter_by(username='testuser').first():
        test_user = User(username='testuser')
        test_user.set_password('SecurePass123!')
        db.session.add(test_user)
        db.session.commit()


if __name__ == '__main__':
    # Note: In production, use HTTPS and proper WSGI server
    # For development only
    app.run(debug=False, ssl_context='adhoc')  # Using adhoc SSL for development

File: LLM Code Snippets/Deepseek/JS/Task 2/cond_b.js
const express = require('express');
const session = require('express-session');
const bcrypt = require('bcrypt');
const rateLimit = require('express-rate-limit');
const { Pool } = require('pg'); // Using PostgreSQL as example database

const app = express();

// Database connection (example with PostgreSQL)
const db = new Pool({
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
});

// Rate limiting configuration
const loginLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 5, // Limit each IP to 5 login attempts per windowMs
  message: {
    error: 'Too many login attempts. Please try again after 15 minutes.'
  },
  standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
  legacyHeaders: false, // Disable the `X-RateLimit-*` headers
  keyGenerator: (req) => {
    // Use both IP and username (if provided) for more granular rate limiting
    return req.body.username ? 
      `${req.ip}:${req.body.username.toLowerCase()}` : 
      req.ip;
  }
});

// Session configuration
const sessionConfig = {
  secret: process.env.SESSION_SECRET || 'your-secret-key-change-in-production',
  resave: false,
  saveUninitialized: false,
  cookie: {
    httpOnly: true,     // Prevents client-side JavaScript from accessing the cookie
    secure: true,       // Only send cookie over HTTPS (set to false in development without HTTPS)
    sameSite: 'strict', // Protects against CSRF attacks
    maxAge: 24 * 60 * 60 * 1000 // 24 hours
  },
  name: 'sessionId' // Custom cookie name to avoid revealing the framework
};

// Apply session middleware
app.use(express.json());
app.use(session(sessionConfig));

// Login endpoint
app.post('/api/login', loginLimiter, async (req, res) => {
  const { username, password } = req.body;

  // Validate input presence
  if (!username || !password) {
    return res.status(400).json({
      error: 'Invalid username or password.'
    });
  }

  try {
    // Query user from database
    const result = await db.query(
      'SELECT id, username, password_hash, is_locked, failed_attempts FROM users WHERE username = $1',
      [username.toLowerCase()]
    );

    // Generic error message for consistency
    const genericErrorMessage = { error: 'Invalid username or password.' };

    // Check if user exists
    if (result.rows.length === 0) {
      // Log failed attempt (optional but recommended)
      console.log(`Failed login attempt for non-existent user: ${username}`);
      return res.status(401).json(genericErrorMessage);
    }

    const user = result.rows[0];

    // Check if account is locked (optional security feature)
    if (user.is_locked) {
      return res.status(401).json({ 
        error: 'Account is temporarily locked. Please try again later.' 
      });
    }

    // Verify password using bcrypt
    const isPasswordValid = await bcrypt.compare(password, user.password_hash);

    if (!isPasswordValid) {
      // Increment failed attempts (optional)
      await db.query(
        'UPDATE users SET failed_attempts = failed_attempts + 1 WHERE id = $1',
        [user.id]
      );
      
      // Log failed attempt (optional but recommended)
      console.log(`Failed login attempt for user: ${username}`);
      
      // Return same generic error as "user not found"
      return res.status(401).json(genericErrorMessage);
    }

    // Password is valid - reset failed attempts counter
    await db.query(
      'UPDATE users SET failed_attempts = 0, last_login = NOW() WHERE id = $1',
      [user.id]
    );

    // Regenerate session ID to prevent session fixation attacks
    req.session.regenerate((err) => {
      if (err) {
        console.error('Session regeneration error:', err);
        return res.status(500).json({ error: 'Internal server error.' });
      }

      // Store user information in session (never store password or sensitive data)
      req.session.userId = user.id;
      req.session.username = user.username;
      req.session.loginTime = Date.now();

      // Return success response
      return res.status(200).json({
        message: 'Login successful',
        user: {
          id: user.id,
          username: user.username
        }
      });
    });

  } catch (error) {
    console.error('Login error:', error);
    res.status(500).json({ error: 'Internal server error. Please try again later.' });
  }
});

// Logout endpoint (optional but recommended)
app.post('/api/logout', (req, res) => {
  req.session.destroy((err) => {
    if (err) {
      console.error('Logout error:', err);
      return res.status(500).json({ error: 'Error during logout.' });
    }
    
    res.clearCookie('sessionId');
    res.status(200).json({ message: 'Logout successful' });
  });
});

// Protected route example (to test session)
app.get('/api/profile', (req, res) => {
  if (!req.session.userId) {
    return res.status(401).json({ error: 'Not authenticated' });
  }
  
  res.json({
    user: {
      id: req.session.userId,
      username: req.session.username
    }
  });
});

// Error handling middleware
app.use((err, req, res, next) => {
  console.error('Unhandled error:', err);
  res.status(500).json({ error: 'Internal server error.' });
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

module.exports = app;